proposals/p2240.md

```diff
@@ -0,0 +1,744 @@
+
+Given that we are only supporting a minimal set of operators, allowing `...` to
+occur in ordinary binary syntax has few advantages and several drawbacks:
+
+-   It might conflict with a future general fold facility.
+-   It would invite users to try other operators, and would probably give less
+    clear errors if they do.
+-   It would substantially complicate parsing and the AST.
+-   It would force users to make a meaningless choice between `x or ...` and
+    `... or x`, and likewise for `and`.
+
+See also the discussion [below](#fold-like-syntax) of using `...,` and `...;` in
+place of the tuple and statement forms of `...`. This is inspired by fold
+expressions, but distinct from them, because `,` and `;` are not truly binary
+operators, and it's targeting a different problem.
+
+### Allow multiple pack expansions in a tuple pattern
+
+As currently proposed, we allow multiple `...` expressions within a tuple
+literal expression, but only allow one `...` pattern within a tuple pattern. It
+is superficially tempting to relax this restriction, but fundamentally
+infeasible.
+
+Allowing multiple `...` patterns would create a potential for ambiguity about
+where their scrutinees begin and end. For example, given a signature like
+`fn F(... each xs: each i32, ... each ys: each i32)`, there is no way to tell
+where `xs` ends and `ys` begins in the argument list; every choice is equally
+valid. That ambiguity can be avoided if the types are different, but that would
+make type _non_-equality a load-bearing part of the pattern. That's a very
+unusual thing to need to reason about in the type system, so it's liable to be a
+source of surprise and confusion for programmers, and in particular it looks
+difficult if not impossible to usefully express with generic types, which would
+greatly limit the usefulness of such a feature.
+
+Furthermore, our solution to the problem of combinatorial explosion in tuple
+pattern matching deeply relies on the fact that there is at most one `...`
+subpattern in the tuple pattern. Without that constraint, it's not at all clear
+if we could type-check variadic pattern matches in polynomial time.
+
+Finally, function authors can straightforwardly work around this restriction by
+adding delimiters. For example, the current design disallows
+`fn F(... each xs: each i32, ... each ys: each i32)`, but it allows
+`fn F((... each xs: each i32), (... each ys: each i32))`, which is not only
+easier to support, but makes the callsite safer and more readable, since the
+boundary between the `xs` and `ys` arguments is explicitly marked. By contrast,
+if we disallowed multiple `...` expressions in a function argument list,
+function callers who ran into that restriction would often find it difficult or
+impossible to work around. Note, however, that this workaround presupposes that
+function signatures can have bindings below top-level, which is currently
+undecided.
```

Comment:
Is there no alternative that doesn't involve destructuring function parameter patterns, where you write `fn F(xs: (... each i32), ys: (... each i32))` or something?